src/client.js

```diff
@@ -237,6 +189,25 @@ class Client extends events.EventEmitter {
       }
     });
   }
+
+  callreclaim() {
+    setTimeout(reclaim = async () => {
+      await this.setConnectionString().catch(err => {});
+      this._timeoutId = setTimeout(reclaim, this._recycleAfter);
+    }, this._recycleAfter);
+  }
```

Comment:
This looks a bit complex!  I think it will not set `this._timeoutId` until the timeout has expired once, for example.  It also assumes that `recycleAfter` is a repeating interval, but it's not -- it's just a one-time thing.

I think there are two pieces to this recycling which, together, will have the desired effect:
 * Make sure we call `client.recycle()` at `credentials.recycleAfter` ms after the last `this.credentials` call.
 * Make sure that we call `this.credentials` if we need a connection string after the `recycleAfter` period has elapsed.
The total effect is this: the Client will recycle at the appropriate time, causing a new connection.  When that new connection goes looking for credentials, the credentials will be out of date, so it will call `this.credentials()` again.

Here's how I'd organize that:
 * have credentials return `recycleAt` (a Date) instead of `recycleAfter`
 * make a `Client.getConnectionString` method which will either return a cached `connectionString` or, if there is no connection string or the time is after `this.recycleAfter`, call `this.credentials`, cache the result, reset the timer, and return it.
 * Make a `Client._setRecycleAt(recycleAt)` that manages `this._recycleTimer`: if any timer exists, clear it, then use `setTimer` to make a new one that will call `this.recycle()` at `recycleAt` (but not loop).

Reply:
I used the `this._timeoutId` to kill the latest timout with`clearTimeout` along with clearing the `recycle` interval when I stop the  `client`. 
 Further you wanted to make client as much independent of method of fetching credentials as possible so I assume that `recycleAfter` may change next time we call pulse service or any other. So it may not be a one time thing.
Previously I think you told to remove the recycleinterval from reclaim function so i did it :)

Reply:
Agreed, `recycleAfter` will be different every time, so an interval (repeating callback) isn't the right choice, but what you've implemented is a similar thing to setInterval, just implemented with setTimeout (that is, the callback for the first timeout calls setTimeout to set the next one..).

Regarding recycleInterval, yes, I think we can get this working without it, then maybe in a follow-up add it back in.

Reply:
If you notice https://github.com/taskcluster/taskcluster-lib-pulse/blob/get-pulse-credential/src/client.js#L286 if there is no `recycleAfter`,  `setConnectionString` will be called only first time (same effect as caching). 
Also `setTimeout` returns id at initiation.
If I try to organize code as you said, then I need to use `recycleAt(date)` to judge whether setConnection was called due to `recycleTimer` and not `recycleInterval` and also use Taskcluster client library to calculate correct time interval since new Date() might be system dependent. 
Once setTimeout is called it will clear by itself once finished. Why to clear it anyway.
I can get connectionString anytime with `this.client.connectionString`
I feel present logic is less complicated :) Don't know if I missed anything

Reply:
I think the correctness issue here is that `callreclaim` sets up a loop of calling `setConnectionString`, but does not actually recycle the client when the connection string changes.  So an existing connection would continue using the old connectionString even after a new connectionString had been fetched.  Also, the first return value from setTimeout in `callreclaim` isn't assigned to anything -- so if that timeout hasn't expired, calling `client.stop` will not actually stop the client.

That said, code can be "correct" (does the necessary things) and still need changes for organizational reasons.  Here I see that the correctness of the overall implementation depends on some careful coordination between the Client and Connection classes.  Connection expects setConnectionString to set client.connectionString as well as (sometimes) client._recycleAfter (an internal attribute).  Then Client relies on Connection to call client.callreclaim, and that method depends on setConnectionString to have set this._recycleAfter in that case (which it probably would have since callreclaim doesn't get called unless the first call to this.credentials returned a recycleAfter, and as we have things implemented now that means the second call to this.credentials will, too -- but that behavior isn't part of the contract for this.credentials!)

Looking at the bigger picture of the coordination of Client and Connection:
 * only the Client ever creates a Connection
 * a Connection only uses its connectionString at connection time -- it never reconnects
 * the Client retires Connections when they should no longer be used

So, Connection shouldn't know anything about this claiming stuff.  It already takes a connectionString and connects to it, and should not need to be changed.  Well, almost -- https://github.com/taskcluster/taskcluster-lib-pulse/blob/get-pulse-credential/src/client.js#L286 is a bit of a misdesign, as really the Client should just be passing a connectionString in to Connection.connect(connectionString), instead of the Connection looking at `this.client.connectionString`.  That was my fault :)

Client needs to do two things:
 * ask for a connectionString just before connecting
 * stop using connections at `recycleAt`

For the first, we can just call `await this.credentials()` right before `newConn.connect(connectionString)`.  We could add some caching to that value (with the result valid until `recycleAt`), using a private Client method like `_fetchCredentials`.  Or, that caching could be implemented in `claimedCredentials`.

For the second, we need to call `this.recycle()` at `recycleAt` -- that's it!  The recycling process will then take care of retiring any connections, setting up a new connection, calling `this.credentials()`, and starting a timer to call `this.recycle()` again.  Managing that timer could be handled in the same `_fetchCredentials` method.

I think a design like this will be more maintainable and grow with the system as we add more functionality.  It's less surprising or confusing to read, which helps the next person successfully change things without introducing bugs.